// @ant-design/graphs/lib/components/flow-graph/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFlowGraphOptions = exports.DEFAULT_OPTIONS = void 0;
const react_1 = __importDefault(require("react"));
const base_1 = require("../../core/base");
const { TextNode } = base_1.RCNode;
exports.DEFAULT_OPTIONS = {
    padding: [20, 0, 0, 50],
    node: {
        type: 'react',
        style: {
            component: (data) => react_1.default.createElement(TextNode, { type: "filled", text: data.id }),
            size: [100, 40],
            ports: [{ placement: 'left' }, { placement: 'right' }],
        },
        state: {
            active: {
                halo: false,
            },
            selected: {
                halo: false,
            },
        },
    },
    edge: {
        type: 'polyline',
        style: {
            lineWidth: 2,
            endArrow: true,
            radius: 8,
            router: { type: 'orth' },
        },
    },
    layout: {
        type: 'dagre',
        rankdir: 'LR',
        animation: false,
    },
    transforms: ['translate-react-node-origin'],
};
const getFlowGraphOptions = ({ direction }) => {
    let options = {};
    if (direction === 'vertical') {
        options = {
            node: {
                style: {
                    ports: [{ placement: 'top' }, { placement: 'bottom' }],
                },
            },
            layout: {
                type: 'dagre',
                rankdir: 'TB',
            },
        };
    }
    return options;
};
exports.getFlowGraphOptions = getFlowGraphOptions;
