// @ant-design/graphs/lib/components/organization-chart/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_OPTIONS = void 0;
exports.getOrganizationChartOptions = getOrganizationChartOptions;
const react_1 = __importDefault(require("react"));
const base_1 = require("../../core/base");
const { TextNode } = base_1.RCNode;
exports.DEFAULT_OPTIONS = {
    padding: [20, 0, 0, 50],
    node: {
        type: 'react',
        style: {
            component: (data) => react_1.default.createElement(TextNode, { type: "filled", text: data.id }),
            size: [100, 40],
            ports: [{ placement: 'top' }, { placement: 'bottom' }],
        },
        state: {
            active: {
                halo: false,
            },
            selected: {
                halo: false,
            },
        },
        animation: false,
    },
    edge: {
        type: 'polyline',
        style: {
            lineWidth: 2,
            router: {
                type: 'orth',
            },
        },
        animation: false,
    },
    layout: {
        type: 'antv-dagre',
    },
    transforms: ['translate-react-node-origin'],
};
function getOrganizationChartOptions({ direction, }) {
    let options = {};
    if (direction === 'vertical') {
        options = {
            node: {
                style: {
                    ports: [{ placement: 'top' }, { placement: 'bottom' }],
                },
            },
            transforms: (prev) => [
                ...prev,
                {
                    type: 'collapse-expand-react-node',
                    key: 'collapse-expand-react-node',
                    enable: false,
                    refreshLayout: true,
                },
            ],
            layout: {
                rankdir: 'TB',
            },
        };
    }
    else {
        options = {
            node: {
                style: {
                    ports: [{ placement: 'left' }, { placement: 'right' }],
                },
            },
            transforms: (prev) => [
                ...prev,
                {
                    type: 'collapse-expand-react-node',
                    key: 'collapse-expand-react-node',
                    iconPlacement: 'right',
                    enable: false,
                    refreshLayout: true,
                },
            ],
            layout: {
                rankdir: 'LR',
            },
        };
    }
    return options;
}
